physics/MP/GFDL/v3_2022/gfdl_cloud_microphys_v3.F90

```diff
@@ -0,0 +1,392 @@
+                  uin, vin, dz, delp, gsize, dtp, hs, water0, rain0,                        &
+                  ice0, snow0, graupel0, hydrostatic, iis, iie, kks, kke, q_con, cappa,     &
+                  fast_mp_consv, adj_vmr, te, dte, prefluxw, prefluxr, prefluxi, prefluxs,  &
+                  prefluxg, last_step, do_inline_mp ) 
+
+      else 
+        write(errmsg,'(*(a))') 'Invalid imp_physics option for GFDL MP v3'
+        errflg = 1
+        return
+      endif 
+      tem   = dtp*con_p001/con_day
+
+      ! fix negative values
+      do i = 1, im
+        !rain0(i)    = max(con_d00, rain0(i))
+        !snow0(i)    = max(con_d00, snow0(i))
+        !ice0(i)     = max(con_d00, ice0(i))
+        !graupel0(i) = max(con_d00, graupel0(i))
+        if(water0(i)*tem < rainmin) then
+          water0(i) = 0.0
+        endif
+        if(rain0(i)*tem < rainmin) then
+          rain0(i) = 0.0
+        endif
+        if(ice0(i)*tem < rainmin) then
+          ice0(i) = 0.0
+        endif
+        if(snow0(i)*tem < rainmin) then
+          snow0(i) = 0.0
+        endif
+        if(graupel0(i)*tem < rainmin) then
+          graupel0(i) = 0.0
+        endif
+      enddo
+
+      ! calculate fraction of frozen precipitation using unscaled
+      ! values of rain0, ice0, snow0, graupel0 (for bit-for-bit)
+      do i=1,im
+        prcp0(i) = (rain0(i)+snow0(i)+ice0(i)+graupel0(i)) * tem
+        if ( prcp0(i) > rainmin ) then
+          sr(i) = (snow0(i) + ice0(i)  + graupel0(i)) &
+                      / (rain0(i) + snow0(i) + ice0(i) + graupel0(i))
+        else
+          sr(i) = 0.0
+        endif
+      enddo
+
+      ! convert rain0, ice0, snow0, graupel0 from mm per day to m per physics timestep
+      water0    = water0*tem
+      rain0    = rain0*tem
+      ice0     = ice0*tem
+      snow0    = snow0*tem
+      graupel0 = graupel0*tem
+
+      ! flip vertical coordinate back
+      do k=1,levs
+        kk = levs-k+1
+        do i=1,im
+
+          if (imp_physics == imp_physics_gfdl) then 
+            gq0(i,k)         = qv1(i,kk)
+            gq0_ntcw(i,k)    = ql1(i,kk) 
+            gq0_ntrw(i,k)    = qr1(i,kk)
+            gq0_ntiw(i,k)    = qi1(i,kk)
+            gq0_ntsw(i,k)    = qs1(i,kk)
+            gq0_ntgl(i,k)    = qg1(i,kk)
+            gq0_ntclamt(i,k) = qa1(i,kk)
+            gt0(i,k)         = pt(i,kk)  
+            gu0(i,k)         = uin(i,kk) 
+            gv0(i,k)         = vin(i,kk) 
+            refl_10cm(i,k)   = refl(i,kk)
+           else 
+            write(errmsg,'(*(a))') 'Invalid imp_physics option for GFDL MP v3' 
+            errflg = 1
+            return
+           endif 
+        enddo
+      enddo
+
+      ! output ice and liquid water 3d precipitation fluxes if requested
+      if (cplchm) then
+        do k=1,levs
+          kk = levs-k+1
+          do i=1,im
+            if (imp_physics==imp_physics_gfdl) then 
```

Comment:
Why is this necessary? Shouldn't the check above already have been executed and returned with an error if imp_physics is wrong?